// @sonoransoftware/sonoran.js/dist/libs/rest/src/lib/REST.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.REST = void 0;
const events_1 = require("events");
const RequestManager_1 = require("./RequestManager");
const constants_1 = require("./utils/constants");
const constants_2 = require("../../../../constants");
const Utils_1 = require("./utils/Utils");
class REST extends events_1.EventEmitter {
    constructor(_instance, _manager, _product, options) {
        super();
        this.instance = _instance;
        this.manager = _manager;
        this.requestManager = new RequestManager_1.RequestManager(_instance, _product, options)
            .on("restDebug" /* Debug */, this.emit.bind(this, "restDebug" /* Debug */))
            .on("rateLimited" /* RateLimited */, this.emit.bind(this, "rateLimited" /* RateLimited */))
            .on("invalidRequestWarning" /* InvalidRequestWarning */, this.emit.bind(this, "invalidRequestWarning" /* InvalidRequestWarning */));
        this.on('newListener', (name, listener) => {
            if (name === "request" /* Request */ || name === "response" /* Response */)
                this.requestManager.on(name, listener);
        });
        this.on('removeListener', (name, listener) => {
            if (name === "request" /* Request */ || name === "response" /* Response */)
                this.requestManager.off(name, listener);
        });
    }
    /**
     * Runs a request from the api
     * @param type API Type Enum
     */
    request(type, ...args) {
        const apiType = constants_1.AllAPITypes.find((aT) => aT.type === type);
        if (!apiType)
            throw new Error('Invalid API Type given for request.');
        let communityId;
        let apiKey;
        switch (apiType.product) {
            case constants_2.productEnums.CAD: {
                communityId = this.instance.cadCommunityId;
                apiKey = this.instance.cadApiKey;
                break;
            }
            case constants_2.productEnums.CMS: {
                communityId = this.instance.cmsCommunityId;
                apiKey = this.instance.cmsApiKey;
                break;
            }
        }
        if (!communityId || !apiKey)
            throw new Error(`Community ID or API Key could not be found for request. P${apiType.product}`);
        if (apiType.minVersion > this.manager.version)
            throw new Error(`[${type}] Subscription version too low for this API type request. Current Version: ${(0, Utils_1.convertSubNumToName)(this.manager.version)} Needed Version: ${(0, Utils_1.convertSubNumToName)(apiType.minVersion)}`);
        const formattedData = this.formatDataArguments(apiType.type, args);
        const options = {
            id: communityId,
            key: apiKey,
            type,
            data: formattedData,
            product: apiType.product
        };
        return this.requestManager.queueRequest(options);
    }
    formatDataArguments(type, args) {
        switch (type) {
            case 'VERIFY_WHITELIST': {
                return {
                    apiId: args[0],
                    accId: constants_2.uuidRegex.test(args[1]) ? args[1] : undefined,
                    serverId: args[2]
                };
            }
            case 'FULL_WHITELIST': {
                return {
                    serverId: args[0]
                };
            }
            case 'GET_COM_ACCOUNT': {
                return {
                    apiId: args[0],
                    username: args[1],
                    accId: args[2],
                    discord: args[3]
                };
            }
            case 'GET_ACCOUNT_RANKS': {
                return {
                    apiId: args[0],
                    username: args[1],
                    accId: args[2],
                    discord: args[3]
                };
            }
            case 'CLOCK_IN_OUT': {
                return {
                    apiId: args[0],
                    accId: args[1],
                    forceClockIn: args[2]
                };
            }
            case 'CHECK_COM_APIID': {
                return {
                    apiId: args[0]
                };
            }
            case 'SET_ACCOUNT_RANKS': {
                return {
                    accountId: args[0],
                    set: args[1],
                    add: args[2],
                    remove: args[3],
                };
            }
            case 'VERIFY_SECRET': {
                return {
                    secret: args[0],
                };
            }
            default: {
                return args;
            }
        }
    }
}
exports.REST = REST;
